### DISCUSSION TITLE
Optimistic updates with optional fields and relationships

### DISCUSSION THREAD

**Discussion by sarink** [urql-graphql/urql#2578] [Q&A] [answered]
As part of making our app offline-first, we're adding optimistic functionality to all mutations. I'm confused how to handle delivering an optimistic update when both the input arguments and the result are dynamic

Consider the following mutation:
```typescript
gql`
  mutation UpdateItem($id: ID!, $name: String, $amount: Int) {
    updateItem(id: $id, name: $name, amount: $amount) {
      id
      name
      amount
    }
  }
`;
```

The consumer may choose to only update the `name` property, but is still asking for the `amount` property in the result. I believe the suggested way to accomplish this is:
```typescript
const graphcache = cacheExchange({
  optimistic: {
    updateItem: (parent, { id, name, amount }, cache) => {
      const existingItem = cache.readFragment(
        gql`
          fragment _ on Item {
            id
            name
            amount
          }
        `,
        { __typename: 'Item', id }
      );

      return {
        __typename: 'Item',
        id,
        name: name ?? existingItem?.name,
        amount: amount ?? existingItem?.amount,
      };
    },
  },
});
```

The API is not super convenient, but so far so good.

Now what happens if a consumer executes _this_ mutation:
```typescript
gql`
  mutation UpdateItem2($id: ID!, $name: String, $amount: Int) {
    updateItem(id: $id, name: $name, amount: $amount) {
      id
      name
      amount
      store {
        id
        itemCount
        warehouse {
          id
          itemCount
        }
      }
      and {
        tons {
          of {
            nested {
              fields {
                wow
              }
            }
          }
        }
      }
    }
  }
`;
```

It seems impossible to write an optimistic updater that Just Works? You'd have to have an intimate (and brittle) knowledge of every field your consumer might ask for, as well as traverse your entire graph with lots of `readFragments` in order to do so.

The optimistic updater page in the [docs](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#optimistic-updates) is very simplistic, and doesn't talk about this problem. Am I misunderstanding something?

Initially, I thought this problem would be easily solved by composing all the consumer's fragments, like:
```typescript
const graphcache = cacheExchange({
  optimistic: {
    updateItem: (parent, { id, name, amount }, cache) => {
      const existingItem = cache.readFragment(
        gql`
          fragment _ on Item {
            ...UpdateItem1Fragment
            ...UpdateItem2Fragment
          }
        `,
        { __typename: 'Item', id }
      );

      return {
        ...existingItem,
        __typename: 'Item',
        id,
        name: name ?? existingItem?.name,
        amount: amount ?? existingItem?.amount,
      };
    },
  },
});
```

However, this also falls apart when you get into modifying relationships, let's expand on our mutation even more and add `storeIds`, which can be used to add/remove the `Item` to different `Store` s:
```typescript
gql`
  mutation UpdateItem3($id: ID!, $name: String, $amount: Int, $storeIds: [ID!]) {
    updateItem(id: $id, name: $name, amount: $amount) {
      id
      name
      amount
      store {
        id
        itemCount
        warehouse {
          id
          itemCount
        }
      }
      and {
        tons {
          of {
            nested {
              fields {
                wow
              }
            }
          }
        }
      }
    }
  }
`;
```

now our optimistic updater looks like:
```
???
```
I'm actually not sure how to write this. I guess you'd have to query every `store` in `storeIds` in its own `readFragment`, ensuring that every consumer who did this also exports that `UpdateItemStoreParts on Store` fragment separately, and build up a list of `stores` ? Then repeat for `and, tons, of, nested, fields, wow` ?

With schema-awareness, the cache already knows all relationships of an `Item`, shouldn't there be a simple way of accessing it? I would have expected this to be a lot easier and written in a way that's universal so it doesn't matter what consumers do or don't do, no need to export and compose fragments or write a 50-line optimistic updater, perhaps an API more like:
```typescript
const graphcache = cacheExchange({
  optimistic: {
    updateItem: (parent, { id, name, amount }, cache) => {
      const existingItemWithAllKnownRelationships = cache.Item.get(id);
      return {
        ...existingItemWithAllKnownRelationships,
        name: name ?? existingItemWithAllKnownRelationships?.name ?? '',
        amount: amount ?? existingItemWithAllKnownRelationships?.amount ?? 0,
      };
    },
  },
});
```

As it stands, I'm not really sure how to go about writing complex optimistic updaters. Any advice?

**Comment by kitten** [comment, marked as answer]
There's a bit of a missing piece here, which I wanted to address a while ago. There should indeed be two things in the optimistic API that we currently don't support:

- We only really are in trouble if we're missing keyable fields or typenames, so we should only error or warn when that's the case
- We should ignore other missing fields that are undefined
- Currently, we look up fields by alias in your optimistic return value. However, we should instead look up fields by name and call functions with the arguments the field has been passed, if we have any

So, long story short, currently you do need a lot of knowledge if you want to return a complete result. However, it's actually completely safe to leave out fields that aren't part of your keyable fields. We just warn you about this field once and then skip it, but it doesn't necessarily break anything

**Comment by sarink** [comment]
@kitten is this true even if the field is marked as nullable? Won't an optimistic response that's missing an already-cached nullable field end up erroneously setting that field to null?

**Reply by kitten** [reply]
No, we always differentiate between undefined and null, and additionally there's a code path for this: https://github.com/FormidableLabs/urql/blob/60388cbf8710e8a25d7030870412c7ae973adaf9/exchanges/graphcache/src/operations/write.ts#L210
